helix-dap/src/client.rs

```diff
@@ -0,0 +1,642 @@
+            .kill_on_drop(true)
+            .spawn();
+
+        let mut process = process?;
+
+        // TODO: do we need bufreader/writer here? or do we use async wrappers on unblock?
+        let writer = BufWriter::new(process.stdin.take().expect("Failed to open stdin"));
+        let reader = BufReader::new(process.stdout.take().expect("Failed to open stdout"));
+
+        Self::streams(
+            Box::new(BufReader::new(reader)),
+            Box::new(writer),
+            id,
+            Some(process),
+        )
+    }
+
+    pub fn id(&self) -> usize {
+        self.id
+    }
+
+    fn next_request_id(&self) -> u64 {
+        self.request_counter.fetch_add(1, Ordering::Relaxed)
+    }
+
+    async fn request(
+        &self,
+        command: String,
+        arguments: Option<Value>,
+        response: bool,
+    ) -> Result<Option<Response>> {
+        let (callback_rx, mut callback_tx) = channel(1);
+
+        let req = Request {
+            back_ch: Some(callback_rx),
+            seq: self.next_request_id(),
+            msg_type: "request".to_owned(),
+            command,
+            arguments,
+        };
+
+        self.server_tx
+            .send(req)
+            .expect("Failed to send request to debugger");
+
+        if response {
+            Ok(Some(callback_tx.recv().await.unwrap()?))
+        } else {
+            Ok(None)
+        }
+    }
+
+    pub fn capabilities(&self) -> &DebuggerCapabilities {
+        self.capabilities
+            .as_ref()
+            .expect("language server not yet initialized!")
+    }
+
+    pub async fn initialize(&mut self, adapter_id: String) -> Result<()> {
+        let args = InitializeArguments {
+            client_id: Some("hx".to_owned()),
+            client_name: Some("helix".to_owned()),
+            adapter_id,
+            locale: Some("en-us".to_owned()),
+            lines_start_at_one: Some(true),
+            columns_start_at_one: Some(true),
+            path_format: Some("path".to_owned()),
+            supports_variable_type: Some(false),
+            supports_variable_paging: Some(false),
+            supports_run_in_terminal_request: Some(false),
+            supports_memory_references: Some(false),
+            supports_progress_reporting: Some(false),
+            supports_invalidated_event: Some(false),
+        };
+
+        let response = self
+            .request("initialize".to_owned(), to_value(args).ok(), true)
+            .await?
+            .unwrap();
+        self.capabilities = from_value(response.body.unwrap()).ok();
+
+        Ok(())
+    }
+
+    pub async fn disconnect(&mut self) -> Result<()> {
+        self.request("disconnect".to_owned(), None, true).await?;
+        Ok(())
+    }
+
+    pub async fn launch(&mut self, args: impl Serialize) -> Result<()> {
+        let mut initialized = self.listen_for_event("initialized".to_owned()).await;
+
+        let res = self.request("launch".to_owned(), to_value(args).ok(), false);
```

Comment:
- [ ] add a quirk flag on this, as typically we want to hear back about launch and Delve does respond

Reply:
Per spec each request needs to have a response yeah. Just the response body might be empty